keps/0036-kudo-operator-toggle-behavior.md

```diff
@@ -0,0 +1,147 @@
+* [Alternatives](#alternatives)
+
+<!-- [Tools for generating]: https://github.com/ekalinin/github-markdown-toc -->
+
+## Summary
+
+This KEP aims to add an ability to switch an operator dependency on and off
+via a specified parameter of a parent operator. Namely, as a result of
+implementing this KEP, `KudoOperator` task definition will get an optional field
+containing a name of a boolean parameter (of the parent operator), the value
+of which will determine whether the child operator instance should exist.
+
+## Motivation
+From [#1775](https://github.com/kudobuilder/kudo/issues/1775):
+> Currently, KudoOperator tasks don't support toggle behavior which
+> significantly complicates the handling of the optional dependencies. A good
+> example could be Spark and History Server or Kafka and Schema Registry. The
+> only possible options to handle it today are either include the templates
+> as-is into the parent operator and control the contents at the template level
+> or to implement a job (via Toggle task) which deploys/removes the dependency
+> based on a parameter value. Also, cascading updates/upgrades become
+> complicated in these scenarios.
+
+### Goals
+
+* Add an ability to enforce existence/**non-existence** of the child
+  operator instance (and all its children) via a parent operator parameter,
+  i.e. to toggle the dependency.
+
+* Make parent operators specifying an invalid parameter for toggling fail
+  instance admission (as opposed to simply failing a plan with misconfigured
+  `KudoOperator` task).
+
+### Non-Goals
+
+* Chain validation of child instances. `KudoOperator` configurations that result
+  in failed instance admission of a child do not fail parent instance admission,
+  only plan execution. Functionality introduced by this proposal is also
+  affected by this: for example, if a parent switches on a child that uses a
+  non-existent parameter for toggling a grandchild, parent instance admission
+  will succeed, only the corresponding parent plan will fail.
+
+* OperatorVersion validation. This proposal adds one more way to create
+  an operator that just cannot be installed regardless of instance parameters.
+
+* Introducing an `UninstallKudoOperator` task. An ability to unconditionally
+  remove a child operator (for example, installed by a previous version), while
+  no more difficult, is not a goal of this proposal.
+
+## Proposal
+
+### Interface
+An optional field `enablingParameter` will be added to the `KudoOperator` task:
+```
+- name: dependency
+  kind: KudoOperator
+  spec:
+    package: "./packages/dependency"
+    parameterFile: dependency-params.yaml
+    enablingParameter: "installDependency"
+```
+In this example, when the parameter `installDependency` equals to `true`, the
+task will be handled as a normal, unconditional `KudoOperator`. When the
+parameter equals to `false`, execution of the task will ensure that
+the corresponding operator instance is not installed (uninstalling it if
+necessary).
```

Comment:
will the uninstall step trigger the clanup plan of the child operator?

Reply:
@akirillov Yes, it will. Normally, removing an instance triggers the cleanup plan (given that the plan exists, hasn't been triggered yet, and so on). In this KEP, there is no plan to treat child instance removal as a special case.